core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java

```diff
@@ -318,28 +134,26 @@ protected void masterOperation(final ClusterAllocationExplainRequest request, fi
         final ShardRouting shardRouting = foundShard;
         logger.debug("explaining the allocation for [{}], found shard [{}]", request, shardRouting);
 
-        getShardStores(shardRouting, new ActionListener<IndicesShardStoresResponse>() {
-            @Override
-            public void onResponse(IndicesShardStoresResponse shardStoreResponse) {
-                ImmutableOpenIntMap<List<IndicesShardStoresResponse.StoreStatus>> shardStatuses =
-                        shardStoreResponse.getStoreStatuses().get(shardRouting.getIndexName());
-                List<IndicesShardStoresResponse.StoreStatus> shardStoreStatus = shardStatuses.get(shardRouting.id());
-                ClusterAllocationExplanation cae = explainShard(shardRouting, allocation, routingNodes,
-                        request.includeYesDecisions(), shardAllocator, shardStoreStatus, gatewayAllocator,
-                        request.includeDiskInfo() ? clusterInfo : null);
-                listener.onResponse(new ClusterAllocationExplainResponse(cae));
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                listener.onFailure(e);
-            }
-        });
+        ClusterAllocationExplanation cae = explainShard(shardRouting, allocation,
+            request.includeDiskInfo() ? clusterInfo : null, request.includeYesDecisions());
+        listener.onResponse(new ClusterAllocationExplainResponse(cae));
     }
 
-    private void getShardStores(ShardRouting shard, final ActionListener<IndicesShardStoresResponse> listener) {
-        IndicesShardStoresRequest request = new IndicesShardStoresRequest(shard.getIndexName());
-        request.shardStatuses("all");
-        shardStoresAction.execute(request, listener);
+    private ClusterAllocationExplanation explainShard(ShardRouting shardRouting, RoutingAllocation allocation,
+                                                      ClusterInfo clusterInfo, boolean includeYesDecisions) {
+        allocation.setDebugMode(includeYesDecisions ? DebugMode.ON : DebugMode.EXCLUDE_YES_DECISIONS);
```

Comment:
I thought about that, but felt that there really is no concrete distinction between "debug" mode and "explain" mode in this case.  Would changing the naming of the enum values help?